mobile/src/main/java/org/openhab/habdroid/ui/PreferencesActivity.kt

```diff
@@ -348,15 +350,46 @@ class PreferencesActivity : AbstractBaseActivity() {
                 ?: preferenceScreen.sharedPreferences.getInt(Constants.PREV_SERVER_FLAGS, 0)
 
             if (flags and ServerProperties.SERVER_FLAG_ICON_FORMAT_SUPPORT == 0) {
-                val iconFormatPreference = preferenceScreen.findPreference(Constants.PREFERENCE_ICON_FORMAT)
                 preferenceScreen.removePreferenceFromHierarchy(iconFormatPreference)
+            } else {
+                iconFormatPreference.setOnPreferenceChangeListener { pref, _ ->
+                    val context = pref.context
+                    clearImageCache(context)
+                    val intent = Intent(context, ItemUpdateWidget::class.java)
+                    intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
+                    val ids = AppWidgetManager.getInstance(context)
+                        .getAppWidgetIds(ComponentName(context, ItemUpdateWidget::class.java))
+                    intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
+                    context.sendBroadcast(intent)
+                    clearImageCache(context)
+                    true
+                }
             }
             if (flags and ServerProperties.SERVER_FLAG_CHART_SCALING_SUPPORT == 0) {
                 val chartScalingPreference = preferenceScreen.findPreference(Constants.PREFERENCE_CHART_SCALING)
                 preferenceScreen.removePreferenceFromHierarchy(chartScalingPreference)
             }
         }
 
+        private fun clearImageCache(context: Context) {
+            // Clear widget icons
+            val files = context.fileList()
+            files.forEach {
+                if (it.matches("^widget-\\d+$".toRegex())) {
+                    Log.d(TAG, "Deleting file $it")
+                    context.deleteFile(it)
+                }
+            }
```

Comment:
I wouldn't use the files directory for storing the icons. I'd suggest to use the cache directory instead, so the cached icons appear as 'cache data' instead of 'app data' in the app listing.
Deleting the cache then essentially becomes something among the lines of `File(context.cacheDir, "widgeticons").listFiles().forEach { f -> f.delete() }`

Reply:
I thought about using the cache directory, but I won't to avoid that any "optimization" app clears the cache and forces us to re-download the icon again, which only wastes data.

Reply:
This argument doesn't make too much sense to me, TBH. It can be equally applied to a web browser's cache, where redownloading web pages wastes data, but still, a web browser's cache lives in the cache directory.

Thinking more about this argument, the whole additional caching layer seems unneeded/redundant. The OkHttp caching layer also caches on disk, so if one passes `FORCE_CACHE_IF_POSSIBLE` as `cachingMode` to `HttpClient.get()`, one achieves exactly the same effect as with the additional layer.

Reply:
How can I get a usable `HttpClient` if `connection` is `null`?

Reply:
That's a very good point. I therefore withdraw the redundancy argument ;-) But please put the cache management code to CacheManager, so we have a single place managing all cache related stuff. I could see two methods there:
- `fun getWidgetIconStream(widgetId: Int): Pair<InputStream, Boolean>?`, where the latter denotes whether the icon is PNG or SVG (think the anyFormat proposal) ... null would be returned if nothing can be found
- `fun saveWidgetIcon(widgetId: Int, iconData: InputStream, isSvg: Boolean)` saves the data to disk.

As for cache vs. files, it's probably fine either way, since I guess it would feel strange if widget icons are gone after clearing the cache directory. We have to make sure the icon for a given widget is cleared when that widget is removed, though. When using the files directory, I'd also prefer if we'd use a subdirectory for the widget images.